### ISSUE TITLE
Improve vimtex compiler and error parsing layer

### ISSUE THREAD

**Issue by lervag** [lervag/vimtex#651]
# Description

There seems to be sufficient desire for using alternative compilers and improved error parsers for me to start work on this. It will require a full refactoring/rewrite of the compiler layer (vimtex/autoload/latexmk.vim). The following should be some good starting points for a todo list:

1. [x] Refactor the latexmk code into a generalized compiler interface and a "latexmk submodule"
2. [x] Separate the compiler code from error parsing
3. [x] Add a proof-of-concept layer for latexrun as an alternative to latexmk

# Some remarks

* This issue replaces/includes #585. Point 2 above should ultimately provide an interface that would allow to use `rubber-info` as a layer between the Vim errorformat and the LaTeX log files. Thanks to @teto for the suggestions and links: [](http://tex.stackexchange.com/questions/10558/use-latexmk-to-filter-the-log-file) and [pplatex](https://github.com/stefanhepp/pplatex).
* This issue replaces #602 (points 1 and 3 above should fix #602).
* This work may/will take some time. Since it is also a major update, I will work on it in a separate branch. I will let (interested) users know about the progress when there is something to update, and I would very much appreciate feedback both on the description of my ideas here and on progress when I have started the work.

**Comment by teto** [comment]
Great initiative !
Maybe ou can get some insipiration from TexStudio which allows to choose different build systems.
I like how vim-grepper https://github.com/mhinz/vim-grepper allows to switch between different search tools with 'Tab'. This might be useful for testing, to easily switch between different tex compilers.

**Comment by clason** [comment]
Excellent idea! For reference, another popular compiler seems to be https://github.com/cereda/arara.

**Comment by languitar** [comment]
Is there any progress on the improved error parsing originating from #585 which I could try out? Or any way to get this going with a hack in the meantime for my document? It's quite bothersome that I cannot go to the location of an error in a huge document like a thesis.

**Comment by languitar** [comment]
Btw, I have just learned that the rubber error parser is also broken and sometimes swallows error messages.

**Comment by lervag** [comment]
I'm sorry, but no, not yet. The reason is that this is a rather big undertaking. Considering that I've had some time lately to work on issues, I might actually start on this in not too long.

I don't think there is any easy hacks either. And I'm not even sure if a successful implementation of the intended feature here will really solve your problem. But I hope so. :)

**Comment by lervag** [comment]
I've started this now. I've done a pretty huge job the last days refactoring and generalizing stuff. The PR #743 is _not_ ready for merge, but it shows you guys where this is heading.

**Comment by lervag** [comment]
Actually, #743 is not that far from ready. I must update the documentation, but that is basically what is missing now.

**Comment by lervag** [comment]
No #743 is ready for merging. I'll wait a couple of days in case anyone is interested in looking at it first. This does provide some pretty substantial changes, but I think it is a pretty good improvement to the code.

**Comment by clason** [comment]
It works for MacVim, but something about the callback is broken with neovim (and neovim-remote).

After compilation, the PDF is shown, but there is no `latexmk compile: success` notification, and the quickfix list doesn't open automatically if there are errors. Entering `\le` does open the list, but also throws an error that is gone too fast to copy -- something like `unknown function: fix_quickfix_paths`.

Everything works on master.

**Comment by lervag** [comment]
Thanks! I've updated and fixed a few problems. It is getting better, and I think the issues you had are solved.

Yes, this PR changes several option names to `g:vimtex_compiler_...`, and the latexmk specific options are now specified in a single Dictionary, `g:vimtex_compiler_latexmk`.

**Comment by clason** [comment]
It's still not working with neovim-remote, but now I get a different issue with `\le` :) "E731: dictionary used as string".

**Comment by lervag** [comment]
Could we move this particular discussion to #743? Thanks! I'm responding more detailed there.

**Comment by clason** [comment]
Sure, sorry!

**Comment by lervag** [comment]
Ok, PR #743 is now merged and I'm ready to address the next part. I'll create a PR for that as well when it's getting ready. It should not be as much work, so I think it will be ready quite soon.

**Comment by lervag** [comment]
Now I've pushed a change that separates the error parsing from the compiler layer. It remains to add a layer for latexrun.

**Comment by clason** [comment]
Seems to work well so far, although there's an inconsistency with the documentation: the `vimtex_quickfix_warnings` option is now `vimtex_qf_latexlog`, not `vimtex_quickfix_latexlog` (as stated in the docs).

**Comment by lervag** [comment]
Thanks for noticing. I've fixed it now.

**Comment by languitar** [comment]
What would need to be done to use pplatex as an error parser now?

**Comment by clason** [comment]
Found another one (sorry!): `latexmk clean` is not working (nor `clean (full)`) on neovim. It reports `finished`, but the files are still there. (It does work on vim8.)

**Comment by lervag** [comment]
Reproduced. I'll work on it asap.

**Comment by lervag** [comment]
It seems there is a bug on `nvim`. If you start it, then try `:!echo "test" &`, it won't do anything. Compare this with the output in Vim.

**Comment by teto** [comment]
It's not really a bug, but one the very few backward incompatibilites with vim: `:!` is disabled. I believe the new way of launching commands is via the api (jobstart etc) or with term, i.e.,
`:term echo "test"`

**Comment by lervag** [comment]
@teto It is not disabled, i.e., doing `:!echo "test"` works as expected. And `:!echo "test" 2>&1 &` works. But `:!echo "test" &` does not. However, I'll consider looking into alternative forms. Perhaps first the `system` function, but later the jobs api.

**Comment by lervag** [comment]
I think #757 should work for most people. I've tested it locally on a new Vim and it works on all cases I've looked at. I'd be happy if someone else could try it out though.

**Comment by clason** [comment]
Continuous compilation works well from my (admittedly short) tests on nvim (master) and MacVim (8.0.525)!

(Minor cosmetic point: if the quickfix list is open and shows errors, and you fix the errors, the list doesn't close anymore after successfully compiling -- but I'm not sure it ever did; maybe I'm misremembering.)

**Comment by lervag** [comment]
@languitar I'm sorry, I forgot to answer your question. To use `pplatex` as a backend for errors, we need to define a new `qf` module, i.e.:

```
qf.vim
qf/latexlog.vim
qf/pplatex.vim [new]
```

`pplatex.vim` should be created similar to `latexlog.vim`. The "object" has to have the following methods: 

1. `init()`: Sets the errorformat (and makeprg, if relevant). Also possibly other initialization.
2. `setqflist(base, jump)`: This is the "main" function. It should parse the log file and generate the quickfix list. See `latexlog.vim` for an example.

Also, of course, `vimtex#qf#...#new()` to return a new instance of the "object".

**Comment by lervag** [comment]
@clason I missed the quickfix thing in the first read. It is fixed now.

I've also pushed an updated latexmk interface that allows to use the new Vim asynchronous stuff. This does not really give a big advantage, except for single shot compilations which become more robust. To use this, set `g:vimtex_compiler_latexmk = {'backend':'jobs'}`. Lacks documentation still, though. And more testing. Also, it doesn't work on windows yet.

**Comment by lervag** [comment]
I will complete this issue by adding a layer for latexrun after #766 is completed.

**Comment by lervag** [comment]
Ok, now #766 is completed. When I get the time I'll begin working on the `latexrun` layer.

**Comment by lervag** [comment]
Now the `latexrun` layer is implemented. I'll update the documentation later.

**Comment by lervag** [comment]
(Note: I have not added quickfix/error parsing for latexrun yet, only compilation.)

**Comment by lervag** [comment]
And now the documentation should also be OK. With that, I consider this issue completed. Time to celebrate :) (Until someone finds that I've added a new bug or that the latexrun backend should be fixed somehow, of course.)

**Comment by languitar** [comment]
@lervag I gave this a try. I think pplatex is actually closer to the `lacheck.vim` file. However, how do I access the path to the logfile when defining the `makeprg`?

**Comment by lervag** [comment]
It's unclear exactly what you've given a try. But, in any case: I think setting the makeprg to the appropriate `pplatex` command, then using something like this, see below, should work:

```vim
" First try something like this
execute 'make' . b:vimtex.log()

" When things work as expected, you should add silent
silent execute ...
```

This should be done in the `s:qf.setqflist(...)` function.

Feel free to open an issue (or PR) if you want to discuss this further.

**Comment by languitar** [comment]
I think I found a solution in #816, which would need a review and some small help with the `errorformat`.

**Comment by lervag** [comment]
Thanks, I'll look into it and help with the efm. Probably some time next
week. :-)

Den lør. 29. apr. 2017, 15.04 skrev Johannes Wienke <
notifications@github.com>:
